### PR TITLE
[mac] simplify scan implementation

### PR DESCRIPTION
**Pull request by abtink**
This commit contains the following changes: It simplifies the MAC
scan implementation by adding common helper  methods such as
`UpdateScanChannel()`.  It also fixes an issue where invoking back-to-
back scan could cause the handler to be overwritten with NULL. It adds
support for special case of Energy Scan with zero scan duration to
collect a single RSSI sample per channel.

### CODE REVIEW

File: src/core/mac/mac.cpp [hunk 1]
```diff
@@ -250,8 +232,54 @@ otError Mac::ConvertBeaconToActiveScanResult(Frame *aBeaconFrame, otActiveScanRe
...
+{
+    otError error = OT_ERROR_NONE;
+
+    do
+    {
+        mScanChannel++;
+        VerifyOrExit(mScanChannel <= OT_RADIO_CHANNEL_MAX, error = OT_ERROR_NOT_FOUND);
+
+    }
+    while ((mScanChannels & (1U << mScanChannels)) == 0);
```
Comment (jwhui) [inline src/core/mac/mac.cpp]: `(1U << mScanChannels)` -> `(1U << mScanChannel)`

File: src/core/mac/mac.cpp [hunk 2]
```diff
@@ -250,8 +232,54 @@ otError Mac::ConvertBeaconToActiveScanResult(Frame *aBeaconFrame, otActiveScanRe
...
+void Mac::PerformEnergyScan(void)
 {
+    otError error = OT_ERROR_NONE;
+
+    if (mScanDuration == 0)
+    {
+        while ((error = UpdateScanChannel()) == OT_ERROR_NONE)
+        {
+            RadioReceive(mScanChannel);
+            ReportEnergyScanResult(otPlatRadioGetRssi(&GetInstance()));
```
Comment (jwhui) [inline src/core/mac/mac.cpp]: These callbacks may occur before `Mac::EnergyScan()` returns.  Do we want to delay them to a following task?

File: src/core/mac/mac.cpp [hunk 3]
```diff
@@ -261,14 +289,15 @@ void Mac::StartEnergyScan(void)
...
-        {
-            // Cancel scan
-            mEnergyScanHandler(mScanContext, NULL);
-            FinishOperation();
-        }
+exit:
+
+    if (error != OT_ERROR_NONE)
+    {
+        mEnergyScanHandler(mScanContext, NULL);
```
Comment (jwhui) [inline src/core/mac/mac.cpp]: See comment above.

File: src/core/mac/mac.cpp (hunk 1 above)
Comment (abtink) [reply src/core/mac/mac.cpp]: Good eye! I am not sure how it got in. I did fix this after testing.

Comment (abtink) [reply src/core/mac/mac.cpp]: Hmmm... checking the PR it looks like when doing "git rebase (squash)" somehow two commits were dropped.

File: src/core/mac/mac.cpp (hunk 2 above)
Comment (abtink) [reply src/core/mac/mac.cpp]: Good point. Will change this to be done from the task instead.

File: src/core/mac/mac.cpp (hunk 3 above)
Comment (abtink) [reply src/core/mac/mac.cpp]: Changed the logic to use the task in the new pushed commit.

**Pull request merged**